### ISSUE TITLE
F.adaptive_avg_pool3d(input, 1) returns infinity in half precision

### ISSUE THREAD

**Issue by jlevyr** [pytorch/pytorch#104602]
### 🐛 Describe the bug

When training a network using half precision, I experience a nan values.
When looking into the network I saw that F.adaptive_avg_pool3d(input, 1) outputs infinity on some inputs.
[bad_input.zip](https://github.com/pytorch/pytorch/files/11950096/bad_input.zip)

It reproduced using different pytorch versions - 1.10-1.12
Also reproduced in different cuda versions - 10-11.7.

code reproduce:
```
import torch
import torch.nn.functional as F

input = torch.load("bad_input.pt").half()
result = F.adaptive_avg_pool3d(bad_input, 1)
print(torch.max(result)) 
```
The print will output infinity.

Possible duplicate issue in https://github.com/pytorch/pytorch/issues/52719.
However, the workadound of x.mean(dim={2,3,4}) does not reproduce the same output as adaptive_avg_pool3d.

### Versions

pip3] numpy==1.22.1
[pip3] pytorch-ranger==0.1.1
[pip3] pytorchvideo==0.1.5
[pip3] torch==1.10.1+cu113
[pip3] torch-optimizer==0.3.0
[pip3] torchmetrics==0.9.3
[pip3] torchvision==0.11.2+cu113

And also here:
pip3] numpy==1.21.5
[pip3] pytorch-ranger==0.1.1
[pip3] pytorchvideo==0.1.5
[pip3] torch==1.13.1+cu117
[pip3] torch-optimizer==0.3.0
[pip3] torchaudio==0.13.1+cu117
[pip3] torchelastic==0.2.0
[pip3] torchmetrics==0.9.3
[pip3] torchtext==0.13.0
[pip3] torchvision==0.14.1+cu117
